File: src/utils/fifoMatcher.ts
import { Position } from '../types'

export interface RawLot {
  ticker: string
  name: string
  qty: number
  price: number
  date: string
  currency: string
  broker: string
  isin?: string
  type: Position['type']
  isSell: boolean
}

/**
 * Given raw buy/sell lots (in any order), applies FIFO matching per ticker.
 * Returns Position[] with sellDate/sellPrice set on closed lots.
 * Partial sells split a buy lot into a closed portion and an open remainder.
 */
export function applyFifo(lots: RawLot[]): Position[] {
  const byTicker: Record<string, RawLot[]> = {}
  for (const lot of lots) {
    (byTicker[lot.ticker] ??= []).push(lot)
  }

  const result: Position[] = []

  for (const tickerLots of Object.values(byTicker)) {
    const buys = tickerLots.filter(l => !l.isSell).sort((a, b) => a.date.localeCompare(b.date))
    const sells = tickerLots.filter(l => l.isSell).sort((a, b) => a.date.localeCompare(b.date))

    if (!sells.length) {
      buys.forEach(b => result.push(toLot(b)))
      continue
    }

    // Mutable queue: each entry tracks remaining qty of a buy lot
    const queue = buys.map(b => ({ ...b, remaining: b.qty }))
    let qi = 0

    for (const sell of sells) {
      let toSell = sell.qty
      while (toSell > 1e-6 && qi < queue.length) {
        const buy = queue[qi]
        const closed = Math.min(buy.remaining, toSell)
        result.push({
          id: '', ticker: buy.ticker, name: buy.name, type: buy.type,
          quantity: closed, buyPrice: buy.price, buyDate: buy.date,
          currency: buy.currency, broker: buy.broker,
          ...(buy.isin ? { isin: buy.isin } : {}),
          sellPrice: sell.price, sellDate: sell.date,
        } as Position)
        buy.remaining -= closed
        toSell -= closed
        if (buy.remaining < 1e-6) qi++
      }
      // Unmatched sell qty → no buy found; silently ignored
    }

    // Remaining open buy lots
    if (qi < queue.length) {
      const first = queue[qi]
      if (first.remaining > 1e-6) {
        result.push({ ...toLot({ ...first, qty: first.remaining }), id: '' } as Position)
      }
      for (let i = qi + 1; i < queue.length; i++) result.push(toLot(queue[i]))
    }
  }

  return result
}

function toLot(l: RawLot): Position {
  return {
    id: '', ticker: l.ticker, name: l.name, type: l.type,
    quantity: l.qty, buyPrice: l.price, buyDate: l.date,
    currency: l.currency, broker: l.broker,
    ...(l.isin ? { isin: l.isin } : {}),
  }
}

File: android/core/calc/src/main/kotlin/com/stocktracker/core/calc/FifoMatcher.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.Position
import com.stocktracker.core.model.PositionType
import java.util.UUID

/** Intermediate struct fed by the import parsers, before FIFO matching. Mirrors RawLot in src/utils/fifoMatcher.ts. */
data class RawLot(
    val ticker: String,
    val name: String,
    val qty: Double,
    val price: Double,
    val date: String,
    val currency: String,
    val broker: String?,
    val isin: String? = null,
    val type: PositionType,
    val isSell: Boolean,
)

private const val EPSILON = 1e-6

private class BuyEntry(val lot: RawLot, var remaining: Double)

/**
 * FIFO lot matcher. Mirrors src/utils/fifoMatcher.ts exactly, including the
 * 1e-6 epsilon that governs partial-lot-consumed boundary behavior, and the
 * silent-drop of any sell quantity left over once the buy queue is exhausted.
 */
fun applyFifo(lots: List<RawLot>): List<Position> {
    val result = mutableListOf<Position>()

    for ((_, group) in lots.groupBy { it.ticker }) {
        val buys = group.filter { !it.isSell }.sortedBy { it.date }
        val sells = group.filter { it.isSell }.sortedBy { it.date }

        if (sells.isEmpty()) {
            buys.forEach { result.add(it.toOpenPosition()) }
            continue
        }

        val queue = buys.map { BuyEntry(it, it.qty) }
        var qi = 0

        for (sell in sells) {
            var toSell = sell.qty
            while (toSell > EPSILON && qi < queue.size) {
                val entry = queue[qi]
                val closedQty = minOf(entry.remaining, toSell)
                result.add(entry.lot.toClosedPosition(closedQty, sell))
                entry.remaining -= closedQty
                toSell -= closedQty
                if (entry.remaining < EPSILON) qi++
            }
            // toSell left over here (buy queue exhausted) is silently dropped, matching the source.
        }

        if (qi < queue.size) {
            val first = queue[qi]
            if (first.remaining > EPSILON) result.add(first.lot.toOpenPosition(quantity = first.remaining))
            for (i in (qi + 1) until queue.size) {
                result.add(queue[i].lot.toOpenPosition())
            }
        }
    }

    return result
}

private fun RawLot.toOpenPosition(quantity: Double = qty): Position = Position(
    id = UUID.randomUUID().toString(),
    ticker = ticker,
    name = name,
    type = type,
    quantity = quantity,
    buyPrice = price,
    buyDate = date,
    currency = currency,
    broker = broker,
    isin = isin,
)

private fun RawLot.toClosedPosition(quantity: Double, sell: RawLot): Position = Position(
    id = UUID.randomUUID().toString(),
    ticker = ticker,
    name = name,
    type = type,
    quantity = quantity,
    buyPrice = price,
    buyDate = date,
    currency = currency,
    broker = broker,
    isin = isin,
    sellPrice = sell.price,
    sellDate = sell.date,
)
